components/layout/query.rs

```diff
@@ -527,21 +530,58 @@ impl FragmentBorderBoxIterator for UnioningFragmentScrollAreaIterator {
 // https://drafts.csswg.org/cssom-view/#extensions-to-the-htmlelement-interface
 impl FragmentBorderBoxIterator for ParentOffsetBorderBoxIterator {
     fn process(&mut self, fragment: &Fragment, level: i32, border_box: &Rect<Au>) {
+        match self.node_offset_box {
+            Some(_) => {  }, // We've already found the node, so we already have its parent.
+            None => {
+                // Remove all nodes at this level or higher, as they can't be parents of this node.
+                self.parent_nodes.truncate(level as usize);
+                assert!(self.parent_nodes.len() == level as usize);
+            },
+        }
+
         if fragment.node == self.node_address {
             // Found the fragment in the flow tree that matches the
             // DOM node being looked for.
-            self.has_found_node = true;
-            self.node_border_box = *border_box;
+            self.has_processed_node = true;
+            self.node_offset_box = Some(NodeOffsetBoxInfo {
+                offset: border_box.origin,
+                rectangle: *border_box,
+            });
 
             // offsetParent returns null if the node is fixed.
             if fragment.style.get_box().position == computed_values::position::T::fixed {
                 self.parent_nodes.clear();
             }
-        } else if level > self.last_level {
+        } else if fragment.contains_node(self.node_address) {
+            // Found a fragment in the flow tree that matches a child
+            // of the DOM node being looked for.
+            // Since we haven't found an exact match, the node must be
+            // inline. Hopefully.
+            match self.node_offset_box {
+                Some(NodeOffsetBoxInfo { ref mut rectangle, .. }) => {
+                    *rectangle = rectangle.union(border_box);
```

Comment:
Okay, Firefox (and Konqueror, remarkably enough) measure `offsetLeft` based on the first line box, which is technically correct, but WebKit/Blink-based browsers just use the first node like this patch currently does. It's the same story for `getClientRects()`, actually&mdash;Gecko correctly returns each whole line's border box, whereas WebKit/Blink split it up into nodes. This should be a separate issue, I suppose.